### PR TITLE
feat: optimize queries with sort and limit by using fixed-size tree chain

### PR DESCRIPTION
**Pull request by bziobrowski**
Work in progress. 
As discussed some time ago - this PR implements limit clause as fixed size tree chain with freelists for both tree and chain nodes . 
It's faster than current tree chain + limit cursor especially for large data sets . That's because it doesn't build tree for all records and only stores smallest/biggest N records. 
All cases except "limit 10, -5" can be optimized this way . 
If proposed approach makes sense I'll add similar LimitedSizeSortedRecordCursor* classes . 

Simple test with console : 

create table trips 
(
    passenger_count int,
    pickup_datetime timestamp,
    trip_distance double
);

insert into trips(passenger_count, pickup_datetime, trip_distance) 
select rnd_int(1,6,0), 
       rnd_timestamp(
        to_timestamp('2015', 'yyyy'),
        to_timestamp('2021', 'yyyy'), 0), 
        rnd_double()*20
from long_sequence(10000000);

select count(*) from trips;  
returns 10 000 000

select trip_distance from trips order by trip_distance desc limit 20 ; 
takes 18.22s without
  and  0.2s    with optimization  

Difference is bigger the more rows there is to sort . 

I applied Vlad's suggestions : 
- moved lo,hi function evaluation to factory of() method so that default SortedLightRecordCursor cursor is used for hi=10,lo=-10 case that can't be optimized and not the new LimitedSizeSortedLightRecordCursor 
- added constant lo,hi handling in SqlCodeGenerator (to detect hi=10,lo=-10 case that can't be optimized)
- for cases that require ignoring rows at start or end of tree chain (e.g.  lo=10,hi=100 or lo=-200,hi=-100 ) I switched from removing these rows from tree to skipping over them in tree cursor (similarly to how it's done in LimitRecordCursor)

I copied some tests and master code fixes from https://github.com/questdb/questdb/pull/1748 to increase coverage .

### CODE REVIEW

File: core/src/main/java/io/questdb/griffin/engine/AbstractRedBlackTree.java [hunk 1]
```diff
@@ -62,13 +64,14 @@ public void close() {
     }
```
Comment (bziobrowski) [inline core/src/main/java/io/questdb/griffin/engine/AbstractRedBlackTree.java:65]: Fixes tree size calculation bug caused by small amounts of memory (smaller than node size) being left at end of each page .

File: core/src/main/java/io/questdb/griffin/engine/orderby/LimitedSizeSortedLightRecordCursor.java [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+     *      L < 0 - last L records
+     * 2. "limit L, H" means we need to keep :
+     *    L < 0          - last  L records (but skip last H records, if H >=0 then don't skip anything    )
+     *    L >= 0, H >= 0 - first H records (but skip first L later, if H <= L then return empty set)
+     *    L >= 0, H < 0  - we can't optimize this case (because it spans from record L-th from the beginning up to
+     *                     H-th from the end, and we don't  ) and need to revert to default behavior -
+     *                     produce the whole set and skip .
+     *
+     *  Similar to LimitRecordCursorFactory.LimitRecordCursor but doesn't check underlying count .
+     */
```
Comment (bziobrowski) [inline core/src/main/java/io/questdb/griffin/engine/orderby/LimitedSizeSortedLightRecordCursor.java:143]: Based on LimitRecordCursor.toTop().

File: core/src/main/java/io/questdb/griffin/engine/orderby/LimitedSizeLongTreeChain.java [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+            long p = root;
+            if (p != -1) {
+                while (leftOf(p) != -1) {
+                    p = leftOf(p);
+                }
+            }
+            chainCurrent = refOf(treeCurrent = p);
+        }
+    }
+
```
Comment (bziobrowski) [inline core/src/main/java/io/questdb/griffin/engine/orderby/LimitedSizeLongTreeChain.java:334]: There are several ways to implement this : 
1. ignoring records in cursor (like in LimitRecordCursor) . Fastest approach if number of records is small and/or cursor won't be traversed many times .  
2. removing first/last N records from tree (implemented here) . It doesn't release any memory though . 
3. same as 2 but without fixing tree structure (could be faster if number of records to skip is large because once result is materialized it probably doesn't need to be a balanced tree anymore ) .

**Comment by puzpuzpuz** [comment]
@bziobrowski recently we've added jdk 17 support to the codebase. I can see compilation errors in CI run on your branch, such as `LimitedSizeSortedLightRecordCursor.java:[21,12] error: reference to Record is ambiguous` (we need to import `*.sql.Record` separately since there is a `java.lang.` class with the same name now). Could you switch to jdk 17 and fix the errors? If you use IntelliJ IDEA, you may need a short instruction from #1674 to get the builds working in IDE.

**Comment by bziobrowski** [comment]
I can't reproduce 
ReaderPoolTest.testGetAndCloseRace:382->assertWithPool:873->assertWithPool:877 [15,12,16,17,18] expected:<3> but was:<6>
failure on Windows .

**Comment by puzpuzpuz** [comment]
> I can't reproduce ReaderPoolTest.testGetAndCloseRace:382->assertWithPool:873->assertWithPool:877 [15,12,16,17,18] expected:<3> but was:<6> failure on Windows .

That's because it's a know flaky test (#1513). I've restarted the build.

**Comment by bluestreak01** [comment]
> 1. Can hi/lo functions be evaluated in SqlCodeGenerator to choose cursor factory (vs of() method ) ?

`hi` and `lo` functions are allowed to be bind variables. When they are constant (not runtime constant!) the `SqlCodeGenerator` can provide choice of factory. When they are bind variables (runtime constants) - factory could provide different cursors depending on values of `hi` and `lo`

File: core/src/test/java/io/questdb/griffin/SortAndLimitTest.java [hunk 4]
```diff
@@ -0,0 +1,195 @@
+package io.questdb.griffin;
+
+import org.junit.Test;
+
+/**
+ *
+ */
+public class SortAndLimitTest extends AbstractGriffinTest {
```
Comment (ideoma) [inline core/src/test/java/io/questdb/griffin/SortAndLimitTest.java:8]: please add tests with `order by DESC`

File: core/src/test/java/io/questdb/griffin/SortAndLimitTest.java [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+                "insert into sorttest select x from long_sequence( 10 );",
+                "insert into sorttest select x from long_sequence( 10 );");
+
+        assertQuery("i\n8\n9\n9\n10\n10\n", "select i from sorttest order by i limit -5");
+    }
+
+    @Test
+    public void testInsertAndSelect_Top_5_returns_last_5_records() throws Exception {
+        runQueries("create table sorttest(i int);",
+                "insert into sorttest select x from long_sequence( 10 );");
```
Comment (ideoma) [inline core/src/test/java/io/questdb/griffin/SortAndLimitTest.java:123]: Would be good if you re-shuffle the records to test that sort works.
Add designated timestamp column and put random values within same partition before the insert, it will randomise the order.

File: core/src/test/java/io/questdb/griffin/SortAndLimitTest.java (hunk 4 above)
Comment (ideoma) [reply core/src/test/java/io/questdb/griffin/SortAndLimitTest.java:8]: please add tests where sorted values are not all distinct

File: core/src/main/java/io/questdb/std/LongList.java [hunk 6]
```diff
@@ -524,4 +524,14 @@ private int scrollUpBlock(int shl, int high, long value) {
         return (high + 1) << shl;
     }
 
-}
\ No newline at end of file
+    public boolean contains(long value) {
```
[the remark this answers is not in the document]
Comment (bziobrowski) [reply core/src/main/java/io/questdb/std/LongList.java]: I removed it because no code was using it anymore .

File: core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java [hunk 7]
```diff
@@ -1201,22 +1202,19 @@ private RecordCursorFactory generateLimit(RecordCursorFactory factory, QueryMode
...
-            final int type = loFunc.getType();
-            if (limitTypes.excludes(type)) {
-                throw SqlException.$(limitLo.position, "invalid type: ").put(ColumnType.nameOf(type));
-            }
-        }
+        return new LimitRecordCursorFactory(factory, loFunc, hiFunc);
+    }
+
+    @Nullable
+    private Function getHiFunction(SqlExecutionContext executionContext, ExpressionNode limitHi) throws SqlException {
```
[the remark this answers is not in the document]
Comment (bziobrowski) [reply core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java]: I reworked both functions to use the same internal function . I didn't get rid of getHiFunction or getLoFunction because there are at least two call sites and inlining method would expose default lo value (that'd be fragile) .

File: core/src/test/java/io/questdb/griffin/SortAndLimitTest.java (hunk 4 above)
Comment (bziobrowski) [reply core/src/test/java/io/questdb/griffin/SortAndLimitTest.java:8]: I copied some tests from https://github.com/questdb/questdb/pull/1748 and also tested tree creation with duplicates .

File: core/src/test/java/io/questdb/griffin/SortAndLimitTest.java (hunk 5 above)
Comment (bziobrowski) [reply core/src/test/java/io/questdb/griffin/SortAndLimitTest.java:123]: I added these kind of tests for both DESC and ASC sort order to SortAndLimiTest .

**Review by bluestreak01 — APPROVED** [review]
